larq/metrics.py

```diff
@@ -0,0 +1,99 @@
+import tensorflow as tf
+from larq import utils
+import numpy as np
+
+
+class MeanChangedValues(tf.keras.metrics.Mean):
+    """Computes the mean ration of changed values in a given tensor.
+
+    !!! example
+        ```python
+        m = metrics.MeanChangedValues()
+        m.update_state(1)
+        m.update_state(2)
+        m.update_state(2)
+        print('Final result: ', m.result().numpy())  # Final result: 0.5
+        ```
+
+    # Arguments
+    values_shape: Shape of the tensor for which to track changes.
+    values_dtype: Data type of the tensor for which to track changes.
+    name: Name of the metric.
+    dtype: Data type of the moving mean.
+    """
+
+    def __init__(
+        self,
+        values_shape=(),
+        values_dtype="int8",
+        name="mean_changed_values",
+        dtype=None,
+    ):
+        super().__init__(name=name, dtype=dtype)
+        self.values_dtype = tf.as_dtype(values_dtype)
+        self.values_shape = tf.TensorShape(values_shape).as_list()
+        self._previous_values = self.add_weight(
+            "previous_values",
+            shape=values_shape,
+            dtype=self.values_dtype,
+            initializer=tf.keras.initializers.zeros,
+        )
+        self._size = np.prod(self.values_shape)
+        self._built = False
+
+    def update_state(self, values, sample_weight=None):
+        values = tf.cast(values, self.values_dtype)
+        if not self._built:
+            self._built = True
+            return self._previous_values.assign(values)
+        changed_values = tf.math.count_nonzero(tf.equal(self._previous_values, values))
+        metric_update_op = super(MeanChangedValues, self).update_state(
+            1 - (tf.cast(changed_values, self.dtype) / self._size)
+        )
+        with tf.control_dependencies([metric_update_op]):
+            return self._previous_values.assign(values)
```

Comment:
Maybe I'm missing something, but I have strong doubts about this. This function returns an op, but it assumes the op is executed only once and `update_state` is called again every time. In eager execution I guess this is fine? The tests follow this pattern, yielding the proper results (but adding nodes to the graph for every `update_state` call in graph mode). In general I would expect this function to create the necessary op which I can then simply execute, i.e. call it only once. I'm not sure how Keras interacts with this.

Perhaps we can initialize `_previous_values` with zeros and ignore the first ratio by using something like `total += ratio*(count > 0)`? This would require implementing the streaming logic ourselves.

Reply:
See also [`keras.metrics.MeanTensor`](https://github.com/tensorflow/tensorflow/blob/3699977134badccb7032fa6921d70e01ba8fdf7d/tensorflow/python/keras/metrics.py#L2349) which also uses a `_built` pattern but avoids returning different ops

Reply:
I had a close look at `keras.metrics.MeanTensor`, because I like that it doesn't require users to initialize the metric with a known shape.
Actually I tried the same approach with lazily calling `add_weight` in `_build` on the first execution of the metric, but that resulted in unnecessary nodes in the graph during eager execution when adding it to a layer. That's a thing I'd like to debug in the future. As for `update_state`, `keras.metrics.MeanTensor` actually uses the same pattern as we (and all other metrics) do here and returns an assign op:
https://github.com/tensorflow/tensorflow/blob/3699977134badccb7032fa6921d70e01ba8fdf7d/tensorflow/python/keras/metrics.py#L2403-L2444

In graph mode, Keras actually sets up a internal conceptual graph and will only add the update ops once. As for eager mode, the metrics is evaluated lazily similar to the pattern you see in the unit tests.

I am confident, that our layer unit tests would fail if we add the metric multiple times.

Reply:
My problem is that unlike any other Keras metric we return two different assign ops, and it is essential they are used both. If keras only creates the ops once (which is indeed what I would expect), it gets the op defined under `if not self._built` and the streaming metric is never updated.

Reply:
You are correct, good catch 👍 

I changed the implementation in 096a6e0